File: app/csharp/AccountingSystem.Domain/Entities/AccountStructure.cs
namespace AccountingSystem.Domain.Entities;

/// <summary>
/// 勘定科目構成エンティティクラス
/// データベースの日本語カラム名と英語プロパティ名をマッピング
/// </summary>
public record AccountStructure
{
    /// <summary>
    /// 勘定科目コード
    /// </summary>
    public string AccountCode { get; init; } = string.Empty;

    /// <summary>
    /// 勘定科目パス（チルダ連結形式: 11~11000~11190~11110）
    /// </summary>
    public string AccountPath { get; init; } = string.Empty;

    /// <summary>
    /// 階層レベル（ルート=1）
    /// </summary>
    public int HierarchyLevel { get; init; }

    /// <summary>
    /// 親科目コード
    /// </summary>
    public string? ParentAccountCode { get; init; }

    /// <summary>
    /// 表示順序（同じ階層内での表示順序）
    /// </summary>
    public int DisplayOrder { get; init; }

    /// <summary>
    /// 作成日時
    /// </summary>
    public DateTime CreatedAt { get; init; }

    /// <summary>
    /// 更新日時
    /// </summary>
    public DateTime UpdatedAt { get; init; }
}

File: app/fsharp/AccountingSystem.Domain/Models/AccountStructure.fs
namespace AccountingSystem.Domain.Models

open AccountingSystem.Domain.Types

/// <summary>
/// 勘定科目構成エンティティ
/// </summary>
type AccountStructure = {
    AccountCode: AccountCode           // 勘定科目コード
    AccountPath: string                // 勘定科目パス（チルダ連結）
    HierarchyLevel: int                // 階層レベル
    ParentAccountCode: AccountCode option  // 親科目コード
    DisplayOrder: int                  // 表示順序
}

/// AccountStructure エンティティのファクトリ関数とユーティリティ
module AccountStructure =
    /// <summary>
    /// 新規 AccountStructure エンティティを作成
    /// </summary>
    let create accountCode accountPath hierarchyLevel displayOrder =
        {
            AccountCode = AccountCode.Create(accountCode)
            AccountPath = accountPath
            HierarchyLevel = hierarchyLevel
            ParentAccountCode = None
            DisplayOrder = displayOrder
        }

    /// <summary>
    /// エンティティ同一性の判定（AccountCode による識別）
    /// </summary>
    let equal (a: AccountStructure) (b: AccountStructure) =
        AccountCode.equal a.AccountCode b.AccountCode

    /// <summary>
    /// エンティティのハッシュコードを取得
    /// </summary>
    let hashCode (structure: AccountStructure) =
        structure.AccountCode.Code.GetHashCode()

    /// <summary>
    /// パスから階層レベルを計算
    /// </summary>
    let calculateLevel (path: string) =
        path.Split('~').Length

    /// <summary>
    /// パスから親科目コードを取得
    /// </summary>
    let getParentCode (path: string) =
        let segments = path.Split('~')
        if segments.Length > 1 then
            Some segments.[segments.Length - 2]
        else
            None

    /// <summary>
    /// 子孫科目かどうかを判定
    /// </summary>
    let isDescendantOf (ancestorCode: AccountCode) (structure: AccountStructure) =
        structure.AccountPath.Contains($"~{ancestorCode.Code}~") ||
        AccountCode.equal structure.AccountCode ancestorCode
